### PR TITLE
providers/google: Add support for Internal Load Balancing

### PR DESCRIPTION
**Pull request by danawillow**
This change enables Internal Load Balancing with the introduction of two new resources: `google_compute_region_backend_service` and `google_compute_health_check`.

For all intents and purposes, this is ready for review. I still need to add documentation.

@evandbrown

### CODE REVIEW

**Comment by cblecker** [comment]
@danawillow: Looks like it needs to be rebased on master, due to vendor updates :)

**Comment by danawillow** [comment]
All right, _now_ it's ready.

File: builtin/providers/google/resource_compute_health_check_test.go [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "google_compute_health_check" {
+			continue
+		}
+
+		_, err := config.clientCompute.HealthChecks.Get(
+			config.Project, rs.Primary.ID).Do()
+		if err == nil {
+			return fmt.Errorf("HealthCheck still exists")
```
Comment (paddyforan) [inline builtin/providers/google/resource_compute_health_check_test.go]: It'd be nice to log the ID of the health check that still exists, to make manual clean-up and debugging easier. :)

File: website/source/docs/providers/google/r/compute_region_backend_service.html.markdown [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+layout: "google"
+page_title: "Google: google_compute_region_backend_service"
+sidebar_current: "docs-google-compute-region-backend-service"
+description: |-
+  Creates a Region Backend Service resource for Google Compute Engine.
+---
+
+# google\_compute\_region\_backend\_service
+
+A Region Backend Service defines a regionally-scoped group of virtual machines that will serve traffic for load balancing.
```
Comment (paddyforan) [inline website/source/docs/providers/google/r/compute_region_backend_service.html.markdown]: A link to the docs would be _awesome_ here!

File: builtin/providers/google/resource_compute_health_check.go [hunk 3]
```diff
@@ -0,0 +1,519 @@
...
+	}
+	if v, ok := d.GetOk("unhealthy_threshold"); ok {
+		hchk.UnhealthyThreshold = int64(v.(int))
+	}
+	if v, ok := d.GetOk("type"); ok {
+		hchk.Type = v.(string)
+	}
+	if v, ok := d.GetOk("tcp_health_check"); ok {
+		if hchk.Type != "TCP" {
+			return fmt.Errorf("TCP health check declared but type is listed as %s", hchk.Type)
```
Comment (danawillow) [inline builtin/providers/google/resource_compute_health_check.go]: Oh neat, great idea! Done.

File: builtin/providers/google/resource_compute_health_check.go [hunk 4]
```diff
@@ -0,0 +1,519 @@
...
+			sslHealthCheck.Response = val.(string)
+		}
+		hchk.SslHealthCheck = sslHealthCheck
+	}
+
+	if v, ok := d.GetOk("http_health_check"); ok {
+		if hchk.Type != "HTTP" {
+			return fmt.Errorf("HTTP health check declared but type is listed as %s", hchk.Type)
+		}
+		httpcheck := v.([]interface{})[0].(map[string]interface{})
```
Comment (danawillow) [inline builtin/providers/google/resource_compute_health_check.go]: I don't totally understand this comment, can you give me an example configuration that could cause this issue?

File: builtin/providers/google/resource_compute_forwarding_rule.go [hunk 5]
```diff
@@ -93,13 +127,24 @@ func resourceComputeForwardingRuleCreate(d *schema.ResourceData, meta interface{
 		return err
 	}
 
+	ps := d.Get("ports").(*schema.Set).List()
```
Comment (paddyforan) [inline builtin/providers/google/resource_compute_forwarding_rule.go]: You're totally right, and I learned some new stuff about `*ResourceData.Get` and `*schema.Set.List` today. Thanks!

File: builtin/providers/google/resource_compute_forwarding_rule.go [hunk 6]
```diff
@@ -93,13 +127,24 @@ func resourceComputeForwardingRuleCreate(d *schema.ResourceData, meta interface{
...
+	}
+
 	frule := &compute.ForwardingRule{
-		IPAddress:   d.Get("ip_address").(string),
-		IPProtocol:  d.Get("ip_protocol").(string),
-		Description: d.Get("description").(string),
-		Name:        d.Get("name").(string),
-		PortRange:   d.Get("port_range").(string),
-		Target:      d.Get("target").(string),
+		BackendService:      d.Get("backend_service").(string),
```
Comment (paddyforan) [inline builtin/providers/google/resource_compute_forwarding_rule.go]: You're totally right, I misunderstood the situations in which `d.Get` would return nil. Whoops! Thanks for correcting me. :D

File: builtin/providers/google/resource_compute_forwarding_rule.go [hunk 7]
```diff
@@ -93,13 +127,24 @@ func resourceComputeForwardingRuleCreate(d *schema.ResourceData, meta interface{
...
+		IPAddress:           d.Get("ip_address").(string),
+		IPProtocol:          d.Get("ip_protocol").(string),
+		Description:         d.Get("description").(string),
+		LoadBalancingScheme: d.Get("load_balancing_scheme").(string),
+		Name:                d.Get("name").(string),
+		Network:             d.Get("network").(string),
```
Comment (paddyforan) [inline builtin/providers/google/resource_compute_forwarding_rule.go]: Yup, you're totally right on this.

File: builtin/providers/google/resource_compute_health_check.go [hunk 8]
```diff
@@ -0,0 +1,519 @@
...
+						"request_path": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+							Default:  "/",
+						},
+					},
+				},
+			},
+
+			"https_health_check": &schema.Schema{
```
Comment (paddyforan) [inline builtin/providers/google/resource_compute_health_check.go]: Hm, these are some good points. My instincts are based on the Go philosophy of not repeating things, but that may be a bad instinct here. Looking at what the `aws_route53_health_check` resource does (as a close equivalent), it uses a single property for all its health checks, and uses "type" to distinguish them, as you previously had. Though that seems to be a reflection of the underlying API.

Thinking this through, I think you're right. It's probably better to stick closer to the API than to try to abstract the repetition out. Good call. :)

File: website/source/docs/providers/google/r/compute_health_check.html.markdown [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+Manages a health check within GCE. This is used to monitor instances
+behind load balancers. Timeouts or HTTP errors cause the instance to be
+removed from the pool. For more information, see [the official
+documentation](https://cloud.google.com/compute/docs/load-balancing/health-checks)
+and
+[API](https://cloud.google.com/compute/docs/reference/latest/healthChecks).
+
+## Example Usage
+
+```js
```
Comment (paddyforan) [inline website/source/docs/providers/google/r/compute_health_check.html.markdown]: You can use `tf` here (I was surprised, too!):

```tf
resource "google_compute_health_check" "default" {
  name = "test"
}
```

I won't comment on all the others, but they can all be updated :)

File: website/source/docs/providers/google/r/compute_health_check.html.markdown [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+resource "google_compute_health_check" "default" {
+  name         = "test"
+
+  timeout_sec        = 1
+  check_interval_sec = 1
+  type               = "TCP"
```
Comment (paddyforan) [inline website/source/docs/providers/google/r/compute_health_check.html.markdown]: Now that this isn't a thing, we should remove it from the example.

File: website/source/docs/providers/google/r/compute_health_check.html.markdown [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+* `tcp_health_check` - (Optional) A TCP Health Check.
+    See *TCP Health Check* below.
+
+* `project` - (Optional) The project in which the resource belongs. If it
+    is not provided, the provider project is used.
+
+* `timeout_sec` - (Optional) The number of seconds to wait before declaring
+    failure (default 5).
+
+* `type` - (Optional) The type of HealthCheck, either TCP, SSL, HTTP or HTTPS
```
Comment (paddyforan) [inline website/source/docs/providers/google/r/compute_health_check.html.markdown]: We should probably remove this, as it no longer exists. :)

File: website/source/docs/providers/google/r/compute_health_check.html.markdown (hunk 9 above)
Comment (danawillow) [inline website/source/docs/providers/google/r/compute_health_check.html.markdown]: Done.

File: website/source/docs/providers/google/r/compute_health_check.html.markdown (hunk 11 above)
Comment (danawillow) [inline website/source/docs/providers/google/r/compute_health_check.html.markdown]: Yes, good call. Done.

**Comment by paddyforan** [comment]
Thanks for all the changes and comments. Taught me a few things. :) I added a second, shorter round of reviews, mostly about documentation and examples. Other than that, I couldn't see any problems. :)

I do notice that we can't run `TestAccComputeForwardingRule_internalLoadBalancing` on our project, probably because it's in beta and we need to be whitelisted. I know @stack72 is working with @evandbrown and others to change the way our GCP tests are run, to take advantage of features like this. Unfortunately, I don't think we can merge this until that happens, as it'll break our tests.

**Comment by paddyforan** [comment]
Ah, you updated the docs, thanks! Took me too long to investigate. :) Running the tests locally, I'm getting this from `make testacc TEST=./builtin/providers/google TESTARGS='-run=TestAccComputeRegionBackendService_*'`

```
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/12/08 18:08:09 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/google -v -run=TestAccComputeRegionBackendService_* -timeout 120m
=== RUN   TestAccComputeRegionBackendService_basic
--- PASS: TestAccComputeRegionBackendService_basic (66.68s)
=== RUN   TestAccComputeRegionBackendService_withBackend
--- FAIL: TestAccComputeRegionBackendService_withBackend (0.01s)
	testing.go:265: Step 0 error: Configuration is invalid.

		Warnings: []string(nil)

		Errors: []string{"google_compute_health_check.default: : invalid or unknown key: type"}
	resource_compute_region_backend_service_test.go:66: Expected TimeoutSec == 10, got 0
	resource_compute_region_backend_service_test.go:69: Expected Protocol to be TCP, got ""
	resource_compute_region_backend_service_test.go:72: Expected 1 backend, got 0
=== RUN   TestAccComputeRegionBackendService_withBackendAndUpdate
--- FAIL: TestAccComputeRegionBackendService_withBackendAndUpdate (0.01s)
	testing.go:265: Step 0 error: Configuration is invalid.

		Warnings: []string(nil)

		Errors: []string{"google_compute_health_check.default: : invalid or unknown key: type"}
	resource_compute_region_backend_service_test.go:107: Expected TimeoutSec == 20, got 0
	resource_compute_region_backend_service_test.go:110: Expected Protocol to be TCP, got ""
	resource_compute_region_backend_service_test.go:113: Expected 1 backend, got 0
FAIL
exit status 1
FAIL	github.com/hashicorp/terraform/builtin/providers/google	66.719s
make: *** [testacc] Error 1
```

Looks like we need to update the tests, too. :( Sorry for the piecemeal review!

**Comment by danawillow** [comment]
Actually, I'm pretty sure the reason you can't run `TestAccComputeForwardingRule_internalLoadBalancing` is because I accidentally broke the test. Give it another try :)

(for what it's worth, ILB is in GA now. Even if it were in beta though, I don't think you would need a whitelist- that's just for alpha [citation needed though].

**Comment by paddyforan** [comment]
Cool! Sorry for the confusion, I couldn't find solid info on that, and when the test failed with ILB stuff, I just assumed. Whoops. I'll re-run the tests and do a final check. :)

**Comment by danawillow** [comment]
No worries!

**Comment by evandbrown** [comment]
Thanks @danawillow and @paddyforan for the hard work on this! Excited to see it in the next release!

**Comment by paddyforan** [comment]
Tests all pass for me. Thanks for sticking with the lengthy review process, and for contributing the features!

**Pull request merged**